New: output block position for screen readers. Update: refactor the way the screen reader information on block button is delivered.

alexstine opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Fixes #24556

- Refactored the block button to be more screen reader accessible.
- Changed the block title to append "Block." to the title for screen reader users.
- Fix the Button component to use the provided id for aria-describedby and the description element.
- Add total blocks announcement.
- Add position announcement out of total.
- Add a message to let screen reader users know the parent block title of child blocks.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->

I tested this using NV Access on Windows 7 using Firefox.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->

Bug Fix and New Feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Commit:
i

Commit:
Try to import getBlockCount.

Commit:
Try to import getBlockCount.

Commit:
Use existing selector.

Commit:
Pass in rootClientId to getBlockCount.

Commit:
Pass our new total arg in to the accessible label function.

Commit:
Attempt to fix tests.

Commit:
Use description text to output instead of aria-label.

Commit:
A fix to button component to use proper ID passed in to span.

Commit:
Define custom unique description ID constructed with clientId of block.

Commit:
Try to output accessible Block label after block title output.

Commit:
Render span instead of div.

Commit:
Get rid of duplicate title. Add spacing instead of : .

Commit:
Auto format.

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

Commit:
Test case.

Commit:
Modify the accessible description function with parent title.

Commit:
Change text.

Commit:
Formatting fixes and additional comments.

alexstine commented on packages/components/src/button/index.js:
```diff
@@ -137,7 +137,7 @@ export function Button( props, ref ) {
 				{ describedBy && (
 					<VisuallyHidden>
-						<span id={ descriptionId }>{ describedBy }</span>
+						<span id={ describedById }>{ describedBy }</span>
```
I was trying to make the aria-describedby id match the id on the element. The way it was before, aria-describedby would take one id and the element would end up with the unique id from lodash. At least that's how the code looked to me, called it a bug fix.

youknowriad replied:
The  way  I understand it is that if  you pass `describedBy` you shouldn't be passing `aria-describedby` to the button because you're telling the "Button" component to render the description itself (os it just generates its own id) while if you pass `aria-describedby` it's the opposite: you render the description outside of the component and you just pass the id.

So I think the right approach here, would be to restore `descriptionId` here and  drop the passed `aria-describedby` 

That said, I agree that it's not clear at first and  it's probably something that should be documented better in the `Button` component README.

alexstine replied:
Do you think the ID will always be unique even with a bunch of blocks? I reverted my change and let it go back to using the randomized ID.

youknowriad replied:
yes, lodash.uniqueId ensure that it's a unique id per component instance.

alexstine commented:
It also seems as if when I ran lint-js, and format-js, it fixed a bunch of files I wasn't working on. Is this normal?

gziolo commented:
> It also seems as if when I ran lint-js, and format-js, it fixed a bunch of files I wasn't working on. Is this normal?

I fixed formatting issues in #29064. You can rebase your branch and formatted files should disappear from the diff.

I also see that you managed to use npm 7 with the repository and that updated the lock file. @ockham and @fluiddot are working on bringing support in #28834. The workaround, for now, is to use `npm install -g npm@6`

youknowriad commented on packages/block-editor/src/components/block-title/index.js:
```diff
@@ -73,5 +74,10 @@ export default function BlockTitle( { clientId } ) {
+	return (
+		<>
+			{ `${ blockInformation.title } ` }
+			<VisuallyHidden as="span">Block.</VisuallyHidden>
```
I think concatenating text like that is local specific and we should probably be using something like: 

```js
sprintf( createInterpolateElement( __( '%s <VisuallyHidden>Block</VisuallyHidden>.' ), { VisuallyHidden: < VisuallyHidden /> } ), title )
```

thought the syntax in this suggestion is not entirely correct, since we'd need interpolation using

youknowriad replied:
Or maybe invert the interpolate and the sprintf call

alexstine replied:
I implemented this.

youknowriad commented on packages/blocks/src/api/utils.js:
```diff
@@ -161,19 +161,25 @@ export function getBlockLabel( blockType, attributes, context = 'visual' ) {
 	const label = getBlockLabel( blockType, attributes, 'accessibility' );
 	const hasPosition = position !== undefined;
+	const parentTitleOutput =
+		parentTitle !== undefined ? ` Child of ${ parentTitle } block.` : '';
```
It seems like the string should be merged with the string  in line 194 to avoid concatenation issues (concatenation is not really allowed in translated strings, we prefer sprintf). which also mean we might need two strings depending on whether there's a parent.

youknowriad replied:
Actually  it seems we need the same logic for all the places that use `parentTitleOutput`

alexstine replied:
Implemented this, it will now use conditionals rather than outputting parentTitleOutput. It's not near as clean in my opinion, but whatever helps me follow coding standards, I'll just have to get used to it.

youknowriad replied:
Thanks :) it's actually not about  coding standards but more about i18n. The translators wouldn't be able to translate the  strings otherwise.

alexstine replied:
Ah I see. That actually makes sense.

alexstine commented:
@gziolo I am now running npm 6 and getting this error while trying to commit.

```
✖ node ./bin/check-latest-npm.js found some errors. Please fix them and try committing again.

Latest NPM check failed!

Error: The local NPM version does not match the expected latest version. Expected 7.5.4, found 6.14.11.

It is required that you have the latest version of NPM installed in order to commit a change to the package-lock.json file.

Run npm install --global npm@latest to install the latest version of NPM. Before retrying your commit, run npm install once more to ensure the package-lock.json contents are correct. If there are any changes to the file, they should be included in your commit.
husky > pre-commit hook failed (add --no-verify to bypass)
```

I will ignore for now, but that's why I ended up on npm 7.

Commit:
No verify.

Commit:
Try to sync package-lock.json with master.

gziolo commented:
> I will ignore for now, but that's why I ended up on npm 7.

Yes, it's unfortunate but that's the way to go.

Commit:
Refresh.

Commit:
Revert package-lock.json to upstream master.

Commit:
Use createInterpolateElement for title output.

Commit:
Fix title block tests.

Commit:
Fix title bugs. Output parentTitle in SR block description via sprintf with conditionals.

alexstine commented:
I have fixed tests for the title block. I see there are some failing tests for the utils.js file, but I'd rather hold off on fixing those until A11Y review. If there's a chance someone finds an issue with what I've done, I'd rather not rewrite all those tests. Title took me minutes to finish, these look a bit more complicated and new tests will have to be added for the new args I added.

While I wait on review, is there any other glaring issues I need to fix?

youknowriad commented on packages/block-editor/src/components/block-list/block-selection-button.js:
```diff
@@ -242,11 +255,22 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 	const blockType = getBlockType( name );
+	let parentTitle;
+	const blockInformation = useBlockDisplayInformation(
+		getBlockRootClientId( clientId )
```
the call to `getBlockRootClientId( clientId )` should be done inside the "useSelect"  callback above. The reason for this is that if the "root" of the clientId changes, the component won't update unless it's used in useSelect.

alexstine replied:
@youknowriad I cannot define the useBlockInformation hook inside a callback, it must be done within a function. ESLint gave me an error when I tried.

alexstine replied:
Oh, you mean I should just get the rootClientId from clientId in the callback? Let me try to do that.

alexstine replied:
Done. That may work, but I haven't been able to test these changes yet. Can't get the build to work.

alexstine replied:
Build worked after npm install again. I gave it a test and it's still working great. I believe passing the rootClientId from the selector is working just fine.

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

Commit:
Remove mods to button component.

Commit:
Pass rootClientId from selector callback.

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg

youknowriad commented on packages/blocks/src/api/utils.js:
```diff
@@ -184,53 +189,78 @@ export function getAccessibleBlockLabel(
-			title,
-			label
-		);
+		return sprintf( '%1$s', label );
```
Seems like  there's no need for the sprintf anymore.

alexstine replied:
Implemented.

youknowriad reviewed (commented):
Looks like there are some e2e test labels to change and the unit tests for the modified function `getAccessibleBlockLabel` are failing for some reason.

Code wise, this is looking good but I'd appreciate some reviews about the logic of these labels from folks that worked on these before. Maybe @talldan or @tellthemachines

Commit:
Remove unused sprintf.

alexstine commented:
@youknowriad I don't plan on fixing those E2E tests until the accessibility team can provide further feedback. If someone feels I'm not going in the right direction, I'd rather not fix all the tests if it's going to change again. Better to wait until we're closer to finished product I think, but all other E2E tests besides the accessible label tests seem to be working.

tikifez commented:
Testing on macOS Catalina using VoiceOver.

For the block selection button the new label of "[label] Block." is showing up. On VoiceOver it looks a little strange at times as there is a comma added sometimes, so "Columns Block., Block navigation structure[…]" so maybe remove the hard stop at the end. 

Where are the 'Column %1$d of %2$d.', et al variants supposed to show up? I've tried blocks which seem like they'd be on (like the obvious columns block) but I don't get them. Since I've tested with and without your changes this is most likely me and not your code.

alexstine commented:
@tikifez Weird. When I look at the markup, all I see is this.

"Columns Block."

That is what I see while in navigation mode. Is it outputting differently on Mac?

Thanks for testing.

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg

tikifez commented:
Here's a screenshot showing what VoiceOver reads. It's pretty verbose and reads that "table 0 columns, 0 rows" in any of the tables in the block editor. Just not as many screen readers for Mac as there are for Windows.

![Screen Shot 2021-02-21 at 11 48 18 AM](https://user-images.githubusercontent.com/1213291/108712943-8608f980-74e5-11eb-9e0b-a29dfa232626.jpg)

alexstine commented:
@tikifez Could you please post specific steps to replicate with Voiceover? I don't have vision, so screenshot doesn't really illustrate the issue for me.

Thanks.

tikifez commented:
Absolutely. Should have to begin with. 

I want to prefix this by saying this is not something I think needs to be fixed, but visually looks odd. Which clearly doesn't have precedence here. But I don't know, which is why i bring it up.

First, a bit about VoiceOver. When VoiceOver is running it shows a floating text box with the text that it is reading aloud. When VoiceOver's settings are set to "high verbosity" or "medium verbosity" the punctuation and letter-casing given don't follow grammatical rules with the label as it stands in this patch. VoiceOver defaults to high verbosity.

What happens is In the document outline on a post or page, VoiceOver adds text to the end of the label. The text is written like the label text doesn't have ending punctuation, adding things like a comma immediately after or additional text in parenthesis. The important bit is that VoiceOver reads it fine. It's only the phrase shown by VoiceOver in the text box which has anything "wrong" with it, with something like "Paragraph Block, period, comma,
 more text" or "Quote Block, period, open parenthesis, text in lowercase, close parenthesis, comma, more text"

To reproduce (if it's even needed), start VoiceFlow with the verbosity set to Medium or High. In WordPress, navigate to a new page or post then tab to the document outline and press space to open, and navigate down the items in the list. When VoiceFlow processes each item in the list it sounds fine, e.g.  "paragraph block, selected block" but shows "Paragraph block, period, open parenthesis, (text in lowercase) selected block, close parenthesis, comma, button.

Now that I've gone on for too long for something which is probably a moot point and I don't feel strongly about, if action is needed, probably removing the period at the end would be fine. Thanks for your patience!

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

alexstine commented:
Hmm, I tested with similar settings on Mac Voiceover and could not replicate. How weird. Maybe there is something I disabled because Voiceover was talking too much for me.

Anyway, I appreciate the report. Let's see if anyone else can figure out how to replicate.

I still have my doubts due to how Voiceover handles the reading of aria-describedby, but at the same time, I do not know a better way to pull this off. :(

Thanks.

tellthemachines commented on packages/block-editor/src/components/block-title/index.js:
```diff
@@ -73,5 +76,14 @@ export default function BlockTitle( { clientId } ) {
+	return createInterpolateElement(
+		sprintf(
+			/* translators: 1: The block title. */
+			__( '%s <VisuallyHidden>Block.</VisuallyHidden>' ),
```
Hiding the word "Block" visually might not work with speech-to-text, because the word is still part of the button label. Assuming speech-to-text software reads labels either from the DOM tree, or from the accessibility API, it won't be able to tell that the word is hidden and may expect users to read out the whole label in order to activate the button.

Would we lose much by dropping the word "block" altogether? Users entering navigation mode should already be focused in the content area, so would expect to be interacting with blocks. Or am I missing something?

alexstine replied:
I added the extra word "block" for 2 reasons.
1. It exists in the current aria-label.
2. I believe it adds context to what the element is. The problem in Gutenberg I feel sometimes is lack of context, this was meant to take away any confusion.

I do understand your point about text to speech software though, this should be tested.

Thanks.

talldan replied:
The reusable block has 'block' in the title so will be read as 'Reusable block Block'. There might be other custom blocks like that too.

Could consider changing it to `<VisuallyHidden>Block:</VisuallyHidden> %s`

It might also be an acceptable trade-off leaving it as it is though, I'll leave it up to you.

guarani replied:
Just a headsup that the block title (which moved here) isn't read on [mobile](https://github.com/WordPress/gutenberg/tree/master/packages/react-native-editor). It looks like <BlockTitle> isn't used they same way on mobile as its text is not read by VoiceOver (I gave this a test on iOS). I haven't got full clarity on what's going on here, but wanted to share.
Aside, it might be best to conditionally on mobile return `blockInformation.title` (as was done previously here) since `createInterpolateElement` seems web-specific.

alexstine replied:
I eliminated title changes from this PR.

tellthemachines commented on packages/blocks/src/api/utils.js:
```diff
@@ -184,53 +189,78 @@ export function getAccessibleBlockLabel(
-				__( '%1$s Block. Column %2$d. %3$s' ),
-				title,
+				/* translators: accessibility text. 1: The block column number. 2: The total number of blocks. 3: The block label.. */
+				__( 'Column %1$d of %2$d. %4$s' ),
```
The last placeholder should be `%3$s` instead of `%4$s`.

alexstine replied:
All fixed, must have missed that one.

Thanks for spotting.

tellthemachines reviewed (commented):
Thanks for working on this! I tested with VoiceOver/Safari (default verbosity settings, I never changed them at all) and the descriptions are read out correctly after the labels, with a slight pause in between. 

The code looks good, apart from a couple of issues I left comments about. Would be good to get this tested by someone experienced with speech-to-text software; is there anyone in the accessibility team that can help out with that?

Commit:
Fix invalid sprintf arg.

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

Commit:
Fix parentTitle output for blocks with label.

talldan commented on packages/block-editor/src/components/block-list/block-selection-button.js:
```diff
@@ -242,11 +259,20 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
 		index + 1,
-		orientation
+		orientation,
+		total,
```
It'd be nice to give this variable a more meaningful name, like 'blockCount'

alexstine replied:
Done.

talldan commented on packages/blocks/src/api/utils.js:
```diff
@@ -161,19 +161,24 @@ export function getBlockLabel( blockType, attributes, context = 'visual' ) {
  *
  * @return {string} The block label.
  */
 export function getAccessibleBlockLabel(
```
I'm a little surprised there aren't unit tests for this function. Probably my fault as I think I initially wrote it.

talldan replied:
Oh, just realise there are unit tests. So these will need to be updated, with new test cases added for the additional branches of code. Happy to help with that, either directly or by providing guidance, let me know.

There are [docs for testing](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/testing-overview.md#javascript-testing), but not sure how good they are.

talldan commented:
@tikifez You seem to be testing List View. This PR modifies the descriptions in Navigation Mode, which can be entered by pressing the 'escape' key when you have a block focused.

talldan reviewed (commented):
Thanks for your contribution @alexstine!

I left a couple of comments, but nothing major. So far its looking good.

There are some end to end tests that need to be fixed—previously these just tested the label which contained the positional information. I think it'd be good to expand these to test the described-by element now so that there's not a loss in test coverage. Is that something you're able to tackle @alexstine? If not I may be able to help with that next week.

talldan commented:
There's another implication of this change I just noticed. `getAccessibleBlockLabel` is used by the react native code base. That this function no longer returns the block title might be an issue for native mobile. I'll ping some folks on the mobile side to see if we can get some feedback.

I think we could also consider renaming getAccessibleBlockLabel to getAccessibleBlockDescription, but happy to make that a separate change so that we can avoid expanding the scope of this.

talldan commented on packages/block-editor/src/components/block-list/block-selection-button.js:
```diff
@@ -262,7 +288,7 @@ function BlockSelectionButton( { clientId, rootClientId, blockElement } ) {
-				label={ label }
+				describedBy={ label }
 			>
 				<BlockTitle clientId={ clientId } />
```
Sorry, one more thing. I think this should always return the proper title of the block, e.g. 'Paragraph'. But some blocks are able to override this using the `__experimentalLabel` function and return a different value. For sighted users this is useful in a few features (List View, breadcrumbs) to discern blocks with the same title, especially where there might be a lot of the same block used as siblings.

The best example of this is the Link block that's used in the Navigation block ([code here](https://github.com/WordPress/gutenberg/blob/aec1ddb17009dcbc82e8b02ee4bb889a148338a2/packages/block-library/src/navigation-link/index.js#L29)). BlockTitle will always return its content (from its label attribute), so if I type 'Home' for a link, BlockTitle returns 'Home'. `getAccessibleLabel` would also output 'Home' through the label value. The actual block title isn't read at all.

For navigation mode it's probably better always have the proper block title here. I don't think there's any advantage to using some of the block content, as this button is directly above the content visually.

alexstine replied:
I believe I removed title changes. Should be okay.

guarani commented on packages/blocks/src/api/utils.js:
```diff
@@ -161,19 +161,24 @@ export function getBlockLabel( blockType, attributes, context = 'visual' ) {
 	attributes,
 	position,
-	direction = 'vertical'
+	direction = 'vertical',
+	total,
+	parentTitle
```
This function is used in a React Native file, which will need updating as well: https://github.com/WordPress/gutenberg/blob/da8555d258edbe676fa079fb51252f918ae032e1/packages/block-editor/src/components/block-list/block.native.js#L191-L195

The new `total` parameter here I understand corresponds to the block count, which comes from: https://github.com/WordPress/gutenberg/blob/da8555d258edbe676fa079fb51252f918ae032e1/packages/block-editor/src/store/selectors.js#L314

Commit:
Give total variable a better descriptive name blocksTotal.

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

Commit:
Merge branch 'master' of https://github.com/wordpress/gutenberg

Commit:
Refresh.

Commit:
Fix conflict.

carolinan commented:
When I apply the PR and try to `run npm build` I get the following error:

SyntaxError: gutenberg\packages\block-editor\src\components\block-list\block-selection-button.js: 
Identifier 'blockInformation' has already been declared (268:7)

I did run npm install first just in case.

Commit:
Merge branch 'trunk' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

Commit:
Correct error that popped up during merge from trunk.

alexstine commented:
@carolinan I fixed it. Looks like a small regression/bug from when I merged trunk in to my branch, it created another variable with the same name conflicting constants. All fixed up now.

Commit:
Revert changes to title.

Commit:
Merge branch 'trunk' of https://github.com/wordpress/gutenberg into new/output-block-position-sr

Commit:
Add title back to description.

alexstine commented:
@talldan Removing the title change seemed to be the best idea for this at least for the moment. Could always re-visit it in the future for a smaller change. I believe what I've done here has addressed most concerns. The only issue still making the tests fail is unit tests. As much as I would like to take on unit tests for this, I just don't have the time to learn about JS unit tests at the moment. Can you help me find someone who may be interested in fixing a few issues with the unit tests?

Much appreciated, we're almost there with this.

I will work on adding the changes to the .native file later this week.

Thanks.

alexstine commented:
So much has changed since I opened this PR. I'm gonna close it out for now and revisit this later down the road. I learned a lot with the first implementation but I think a bigger problem needs to be tackled first.
